src/Decompiler/Scanning/VarargsFormatScanner.cs

```diff
@@ -0,0 +1,189 @@
+﻿#region License
+/* 
+ * Copyright (C) 1999-2016 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Reko.Core;
+using Reko.Core.CLanguage;
+using Reko.Core.Code;
+using Reko.Core.Expressions;
+using Reko.Core.Serialization;
+using Reko.Core.Types;
+using Reko.Evaluation;
+
+namespace Reko.Scanning
+{
+    /// <summary>
+    /// Try to read varargs format, then to parse it, get arguments types and
+    /// build application instruction.
+    /// </summary>
+    class VarargsFormatScanner
+    {
+        private Program program;
+        private IProcessorArchitecture arch;
+        private Frame frame;
+        private ExpressionSimplifier eval;
+        private FunctionType sig;
+
+        public VarargsFormatScanner(
+            Program program,
+            Frame frame,
+            EvaluationContext ctx)
+        {
+            this.program = program;
+            this.arch = program.Architecture;
+            this.frame = frame;
+            this.eval = new ExpressionSimplifier(ctx);
+        }
+
+        public Instruction BuildInstruction(Expression callee, CallSite site)
+        {
```

Comment:
I'd love to see some unit tests for `BuildInstruction` that exercise it thoroughly.